src/v/cloud_storage/remote.h

```diff
@@ -21,8 +21,26 @@
 #include <seastar/core/gate.hh>
 #include <seastar/core/loop.hh>
 
+#include <utility>
+
 namespace cloud_storage {
 
+/// \brief Predicate required to continue operation
+///
+/// Describes a predicate to be evaluated before starting an expensive
+/// operation, or to be evaluated when an operation has failed, to find
+/// the reason for failure
+struct should_continue_operation {
+    ss::sstring stop_condition;
+    std::function<bool()> predicate;
+    should_continue_operation(
+      ss::sstring description, const std::function<bool()>& predicate)
+      : stop_condition(std::move(description))
+      , predicate(predicate) {}
```

Comment:
nit: this is fine, but since we are copying the std::function anyway, it's a bit cleaner to have the caller do the copy: 

```
      ss::sstring description, std::function<bool()> predicate)
      : stop_condition(std::move(description))
      , predicate(std::move(predicate)) {}
```

Reply:
You can also replace `std::function` with `ss::noncopyable_function` from Seastar.

Reply:
fixed now as more reviews are required, and used `ss:noncopyable_function` in place of `std::function`